File: helper/CodexUsageAlert.m
#import <AppKit/AppKit.h>

@interface AlertController : NSObject <NSApplicationDelegate, NSWindowDelegate>
@property(nonatomic, strong) NSDictionary<NSString *, NSString *> *options;
@property(nonatomic, strong) NSPanel *panel;
@property(nonatomic, strong) NSTimer *timer;
@end

@implementation AlertController

- (instancetype)initWithOptions:(NSDictionary<NSString *, NSString *> *)options {
    self = [super init];
    if (self) _options = options;
    return self;
}

- (NSColor *)accentColor {
    NSString *level = self.options[@"--level"];
    if ([level isEqualToString:@"critical"]) return NSColor.systemRedColor;
    if ([level isEqualToString:@"severe"]) return NSColor.systemOrangeColor;
    return NSColor.systemBlueColor;
}

- (void)applicationDidFinishLaunching:(NSNotification *)notification {
    NSString *preferredLanguage = NSLocale.preferredLanguages.firstObject ?: @"en";
    BOOL isChinese = [preferredLanguage.lowercaseString hasPrefix:@"zh"];
    NSString *titleText = self.options[isChinese ? @"--title-zh" : @"--title-en"];
    NSString *messageText = self.options[isChinese ? @"--message-zh" : @"--message-en"];
    NSRect rect = NSMakeRect(0, 0, 420, 168);
    self.panel = [[NSPanel alloc] initWithContentRect:rect
                                            styleMask:NSWindowStyleMaskTitled |
                                                      NSWindowStyleMaskClosable |
                                                      NSWindowStyleMaskFullSizeContentView |
                                                      NSWindowStyleMaskNonactivatingPanel
                                              backing:NSBackingStoreBuffered
                                                defer:NO];
    self.panel.titleVisibility = NSWindowTitleHidden;
    self.panel.titlebarAppearsTransparent = YES;
    self.panel.movableByWindowBackground = YES;
    self.panel.releasedWhenClosed = NO;
    self.panel.hidesOnDeactivate = NO;
    self.panel.level = NSFloatingWindowLevel;
    self.panel.collectionBehavior = NSWindowCollectionBehaviorCanJoinAllSpaces |
                                    NSWindowCollectionBehaviorFullScreenAuxiliary;
    self.panel.delegate = self;
    [self.panel standardWindowButton:NSWindowMiniaturizeButton].hidden = YES;
    [self.panel standardWindowButton:NSWindowZoomButton].hidden = YES;

    NSVisualEffectView *background = [[NSVisualEffectView alloc] initWithFrame:rect];
    background.material = NSVisualEffectMaterialHUDWindow;
    background.blendingMode = NSVisualEffectBlendingModeBehindWindow;
    background.state = NSVisualEffectStateActive;
    background.wantsLayer = YES;
    background.layer.cornerRadius = 14;
    background.layer.borderWidth = 2;
    background.layer.borderColor = self.accentColor.CGColor;
    self.panel.contentView = background;

    NSTextField *title = [NSTextField labelWithString:titleText];
    title.font = [NSFont systemFontOfSize:17 weight:NSFontWeightSemibold];
    title.textColor = self.accentColor;

    NSTextField *message = [NSTextField wrappingLabelWithString:messageText];
    message.font = [NSFont systemFontOfSize:13];
    message.maximumNumberOfLines = 3;
    message.lineBreakMode = NSLineBreakByWordWrapping;

    NSTimeInterval duration = [self.options[@"--duration"] doubleValue];
    NSString *timingText;
    if (duration > 0) {
        timingText = isChinese
            ? [NSString stringWithFormat:@"%ld 秒后自动关闭", (long)duration]
            : [NSString stringWithFormat:@"Closes automatically in %ld seconds", (long)duration];
    } else {
        timingText = isChinese
            ? @"请手动关闭此紧急提醒"
            : @"Please dismiss this critical alert manually";
    }
    NSTextField *timing = [NSTextField labelWithString:timingText];
    timing.font = [NSFont systemFontOfSize:11];
    timing.textColor = NSColor.secondaryLabelColor;

    NSButton *closeButton = [NSButton buttonWithTitle:(isChinese ? @"关闭" : @"Dismiss")
                                               target:self
                                               action:@selector(closeAlert:)];
    closeButton.bezelStyle = NSBezelStyleRounded;
    closeButton.keyEquivalent = @"\033";

    for (NSView *view in @[title, message, timing, closeButton]) {
        view.translatesAutoresizingMaskIntoConstraints = NO;
        [background addSubview:view];
    }
    [NSLayoutConstraint activateConstraints:@[
        [title.leadingAnchor constraintEqualToAnchor:background.leadingAnchor constant:22],
        [title.trailingAnchor constraintLessThanOrEqualToAnchor:closeButton.leadingAnchor constant:-16],
        [title.topAnchor constraintEqualToAnchor:background.topAnchor constant:27],
        [closeButton.trailingAnchor constraintEqualToAnchor:background.trailingAnchor constant:-20],
        [closeButton.centerYAnchor constraintEqualToAnchor:title.centerYAnchor],
        [message.leadingAnchor constraintEqualToAnchor:title.leadingAnchor],
        [message.trailingAnchor constraintEqualToAnchor:background.trailingAnchor constant:-22],
        [message.topAnchor constraintEqualToAnchor:title.bottomAnchor constant:13],
        [timing.leadingAnchor constraintEqualToAnchor:title.leadingAnchor],
        [timing.bottomAnchor constraintEqualToAnchor:background.bottomAnchor constant:-19],
    ]];

    NSScreen *screen = NSScreen.mainScreen ?: NSScreen.screens.firstObject;
    if (screen) {
        NSRect frame = screen.visibleFrame;
        [self.panel setFrameOrigin:NSMakePoint(
            NSMaxX(frame) - NSWidth(self.panel.frame) - 24,
            NSMaxY(frame) - NSHeight(self.panel.frame) - 24
        )];
    } else {
        [self.panel center];
    }
    [self.panel orderFrontRegardless];

    if (duration > 0) {
        self.timer = [NSTimer scheduledTimerWithTimeInterval:duration
                                                     target:self
                                                   selector:@selector(closeAlert:)
                                                   userInfo:nil
                                                    repeats:NO];
    }
}

- (void)closeAlert:(id)sender {
    [self.timer invalidate];
    [self.panel close];
    [NSApp terminate:nil];
}

- (void)windowWillClose:(NSNotification *)notification {
    [self.timer invalidate];
    [NSApp terminate:nil];
}

@end

static NSDictionary<NSString *, NSString *> *ParseOptions(int argc, const char *argv[]) {
    NSMutableDictionary *values = [NSMutableDictionary dictionary];
    for (int index = 1; index + 1 < argc; index += 2) {
        NSString *key = [NSString stringWithUTF8String:argv[index]];
        NSString *value = [NSString stringWithUTF8String:argv[index + 1]];
        if ([key hasPrefix:@"--"] && value) values[key] = value;
    }
    if (!values[@"--title-zh"] || !values[@"--message-zh"] ||
        !values[@"--title-en"] || !values[@"--message-en"] ||
        !values[@"--duration"]) return nil;
    return values;
}

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        NSDictionary *options = ParseOptions(argc, argv);
        if (!options) {
            fprintf(stderr, "invalid alert arguments\n");
            return 2;
        }
        NSApplication *application = NSApplication.sharedApplication;
        AlertController *controller = [[AlertController alloc] initWithOptions:options];
        [application setActivationPolicy:NSApplicationActivationPolicyAccessory];
        application.delegate = controller;
        [application run];
    }
    return 0;
}

File: helper/CodexUsageAlert.swift
import AppKit

private struct AlertOptions {
    let level: String
    let title: String
    let message: String
    let duration: TimeInterval
    let isChinese: Bool

    static func parse(_ arguments: [String]) -> AlertOptions? {
        var values: [String: String] = [:]
        var index = 1
        while index + 1 < arguments.count {
            if arguments[index].hasPrefix("--") {
                values[arguments[index]] = arguments[index + 1]
                index += 2
            } else {
                index += 1
            }
        }
        let isChinese = (Locale.preferredLanguages.first ?? "en")
            .lowercased().hasPrefix("zh")
        let titleKey = isChinese ? "--title-zh" : "--title-en"
        let messageKey = isChinese ? "--message-zh" : "--message-en"
        guard let title = values[titleKey],
              let message = values[messageKey],
              let durationText = values["--duration"],
              let duration = TimeInterval(durationText), duration >= 0 else {
            return nil
        }
        return AlertOptions(
            level: values["--level"] ?? "ok",
            title: title,
            message: message,
            duration: duration,
            isChinese: isChinese
        )
    }
}

@MainActor
private final class AlertController: NSObject, NSApplicationDelegate, NSWindowDelegate {
    private let options: AlertOptions
    private var panel: NSPanel?
    private var timer: Timer?

    init(options: AlertOptions) {
        self.options = options
    }

    func applicationDidFinishLaunching(_ notification: Notification) {
        let panel = NSPanel(
            contentRect: NSRect(x: 0, y: 0, width: 420, height: 168),
            styleMask: [.titled, .closable, .fullSizeContentView, .nonactivatingPanel],
            backing: .buffered,
            defer: false
        )
        panel.titleVisibility = .hidden
        panel.titlebarAppearsTransparent = true
        panel.isMovableByWindowBackground = true
        panel.isReleasedWhenClosed = false
        panel.hidesOnDeactivate = false
        panel.level = .floating
        panel.collectionBehavior = [.canJoinAllSpaces, .fullScreenAuxiliary]
        panel.delegate = self
        panel.standardWindowButton(.miniaturizeButton)?.isHidden = true
        panel.standardWindowButton(.zoomButton)?.isHidden = true

        let background = NSVisualEffectView()
        background.material = .hudWindow
        background.blendingMode = .behindWindow
        background.state = .active
        background.wantsLayer = true
        background.layer?.cornerRadius = 14
        background.layer?.borderWidth = 2
        background.layer?.borderColor = accentColor.cgColor
        panel.contentView = background

        let title = NSTextField(labelWithString: options.title)
        title.font = .systemFont(ofSize: 17, weight: .semibold)
        title.textColor = accentColor

        let message = NSTextField(wrappingLabelWithString: options.message)
        message.font = .systemFont(ofSize: 13)
        message.maximumNumberOfLines = 3
        message.lineBreakMode = .byWordWrapping

        let timing = NSTextField(labelWithString: options.duration > 0
            ? (options.isChinese
                ? "\(Int(options.duration)) 秒后自动关闭"
                : "Closes automatically in \(Int(options.duration)) seconds")
            : (options.isChinese
                ? "请手动关闭此紧急提醒"
                : "Please dismiss this critical alert manually"))
        timing.font = .systemFont(ofSize: 11)
        timing.textColor = .secondaryLabelColor

        let closeButton = NSButton(
            title: options.isChinese ? "关闭" : "Dismiss",
            target: self,
            action: #selector(closeAlert)
        )
        closeButton.bezelStyle = .rounded
        closeButton.keyEquivalent = "\u{1b}"

        for view in [title, message, timing, closeButton] {
            view.translatesAutoresizingMaskIntoConstraints = false
            background.addSubview(view)
        }

        NSLayoutConstraint.activate([
            title.leadingAnchor.constraint(equalTo: background.leadingAnchor, constant: 22),
            title.trailingAnchor.constraint(lessThanOrEqualTo: closeButton.leadingAnchor, constant: -16),
            title.topAnchor.constraint(equalTo: background.topAnchor, constant: 27),

            closeButton.trailingAnchor.constraint(equalTo: background.trailingAnchor, constant: -20),
            closeButton.centerYAnchor.constraint(equalTo: title.centerYAnchor),

            message.leadingAnchor.constraint(equalTo: title.leadingAnchor),
            message.trailingAnchor.constraint(equalTo: background.trailingAnchor, constant: -22),
            message.topAnchor.constraint(equalTo: title.bottomAnchor, constant: 13),

            timing.leadingAnchor.constraint(equalTo: title.leadingAnchor),
            timing.bottomAnchor.constraint(equalTo: background.bottomAnchor, constant: -19),
        ])

        if let screen = NSScreen.main ?? NSScreen.screens.first {
            let frame = screen.visibleFrame
            panel.setFrameOrigin(NSPoint(
                x: frame.maxX - panel.frame.width - 24,
                y: frame.maxY - panel.frame.height - 24
            ))
        } else {
            panel.center()
        }

        self.panel = panel
        panel.orderFrontRegardless()

        if options.duration > 0 {
            timer = Timer.scheduledTimer(withTimeInterval: options.duration, repeats: false) { _ in
                Task { @MainActor in self.closeAlert() }
            }
        }
    }

    private var accentColor: NSColor {
        switch options.level {
        case "critical": return .systemRed
        case "severe": return .systemOrange
        default: return .systemBlue
        }
    }

    @objc private func closeAlert() {
        timer?.invalidate()
        panel?.close()
        NSApplication.shared.terminate(nil)
    }

    func windowWillClose(_ notification: Notification) {
        timer?.invalidate()
        NSApplication.shared.terminate(nil)
    }
}

guard let options = AlertOptions.parse(CommandLine.arguments) else {
    FileHandle.standardError.write(Data("invalid alert arguments\n".utf8))
    exit(2)
}

let application = NSApplication.shared
let controller = AlertController(options: options)
application.setActivationPolicy(.accessory)
application.delegate = controller
application.run()
